bot/exts/valentines/markov_poem_generator.py

```diff
@@ -0,0 +1,317 @@
+        scheme = self.templates.get(scheme, scheme)
+        self.scheme = scheme
+
+        unit_count = self._get_unit_count(scheme)
+
+        time_start = datetime.now()
+
+        try:
+            async with async_timeout.timeout(self.POEM_TIMEOUT), ctx.typing():
+                stanzas = []
+                acc_lines = []  # Accumulate lines before joining into a stanza
+                rhyme_track = {}  # Maps units to their rhyme sets
+
+                for unit in scheme:
+                    is_last_line = True if unit_count[unit] == 1 else False
+
+                    # Create new stanza
+                    if unit == "/":
+                        new_stanza = "\n".join(acc_lines)
+                        stanzas.append(new_stanza)
+                        acc_lines = []
+                        continue
+
+                    # Creating a line for the unit
+                    if unit not in rhyme_track:
+                        new_line = await self._init_unit(
+                            is_last_line, rhyme_track, unit
+                        )
+                        acc_lines.append(new_line)
+                    else:
+                        new_line = await self._get_rhyming_line(
+                            word_rhymes=rhyme_track[unit]
+                        )
+                        acc_lines.append(new_line)
+
+                        # If last line, it will not be referred to again
+                        if not is_last_line:
+                            rhyme_track[unit] |= await self._get_rhyme_set(
+                                instance=self,
+                                word=self._get_last_word(new_line),
+                                is_instant_fail=(not is_last_line)
+                            )
+
+                    unit_count[unit] -= 1
+
+                stanzas.append("\n".join(acc_lines))  # Append final stanza
+
+                elapsed_time = datetime.now() - time_start
+                elapsed_time = elapsed_time.seconds
+
+                poem_embed = Embed(
+                    title="A Markov Poem For " + str(ctx.author.name),
+                    color=Colours.pink,
+                    description="\n\n".join(stanzas)
+                )
+                poem_embed.set_footer(text=f"Elapsed time: {elapsed_time}s\n"
+                                      "Rhymes API provided by datamuse.")
+                await ctx.send(embed=poem_embed)
+        except TimeoutError:
+            logging.warning("Poem generator timed out.")
+            await ctx.send("Unlucky, try again!")
+
+    async def cog_command_error(
+        self,
+        ctx: commands.Context,
+        error: Exception
+    ) -> None:
+        """Handles Discord errors and exceptions."""
+        if isinstance(error, RhymingSentenceNotFound):
+            logging.warning("Rhyming sentence not found!")
+            await ctx.send("Sentence failed, trying again...")
+            return await self.poem(ctx, self.scheme)
```

Comment:
Currently, using `async_timeout`. Once the time runs out, it will just return an error, and it will not try again. This is not the most effective way of dealing with it. It is one of the things that needs a more effective way.

Reply:
No what i mean is:
```md
-> users sends requests for poem
-> gets RhymingSentenceNotFound
-> tries again for with the scheme
-> gets RhymingSentenceNotFound
... This would be a infinite loop
```

Reply:
I know, but it won't go on forever as the timeout will kill the loop if it goes on beyond a certain number of seconds.

Reply:
Alright my bad.